lib/Dialect/FIRRTL/Transforms/IMDCE.cpp

```diff
@@ -0,0 +1,412 @@
+//===- IMDCE.cpp - Inter Module Dead Code Elimination -----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-imdce"
+
+using namespace circt;
+using namespace firrtl;
+
+/// Return true if this is a wire or a register or a node.
+static bool isWireOrRegOrNode(Operation *op) {
+  return isa<WireOp, RegResetOp, RegOp, NodeOp>(op);
+}
+
+/// Return true if this is a wire or register we're allowed to delete.
+static bool isDeletableWireOrRegOrNode(Operation *op) {
+  if (auto name = dyn_cast<FNamableOp>(op))
+    if (!name.hasDroppableName())
+      return false;
+  return !hasDontTouch(op);
+}
+
+namespace {
+struct IMDCEPass : public IMDCEBase<IMDCEPass> {
+  void runOnOperation() override;
+
+  void rewriteModuleSignature(FModuleOp module);
+  void rewriteModuleBody(FModuleOp module);
+
+  void markAlive(Value value) {
+    // If the value is already in `liveSet`, skip it.
+    if (liveSet.count(value))
+      return;
+    liveSet.insert(value);
+    worklist.push_back(value);
+  }
+
+  /// Return true if the value is known alive.
+  bool isKnownAlive(Value value) const {
+    assert(value && "null should not be used");
+    return liveSet.count(value);
+  }
+
+  /// Return true if the value is assumed dead.
+  bool isAssumedDead(Value value) const { return !isKnownAlive(value); }
+
+  bool isBlockExecutable(Block *block) const {
+    return executableBlocks.count(block);
+  }
+  void visitUser(Operation *op);
+  void visitValue(Value value);
+  void visitConnect(FConnectLike connect);
+  void visitSubelement(Operation *op);
+  void markBlockExecutable(Block *block);
+  void markWireOrRegOrNode(Operation *op);
+  void markMemOp(MemOp op);
+  void markInstanceOp(InstanceOp instanceOp);
+  void markUnknownSideEffectOp(Operation *op);
+
+private:
+  /// The set of blocks that are known to execute, or are intrinsically alive.
+  SmallPtrSet<Block *, 16> executableBlocks;
```

Comment:
There is at least 1 block per module, there are probably more than 16 modules.  I think it would be better if this was a `DenseSet`.

Reply:
I agree. Probably it would be good to update IMCP as well.